File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/ViewModel/TheBottom/FungiBeast_ViewModel.ts
import { Spine } from "csharp";
import { Mathf } from "../../../../../../../Core/Module/Math/Mathf";
import { ApplyPowerAction } from "../../../../Action/Common/ApplyPowerAction";
import DamageAction from "../../../../Action/Common/DamageAction";
import { DoFuncAction } from "../../../../Action/Common/DoFuncAction";
import { RollMoveAction } from "../../../../Action/Common/RollMoveAction";
import { AttackEffect } from "../../../../DataDefine/AttackEffect";
import DamageInfo from "../../../../DataDefine/DamageInfo";
import { DamageType } from "../../../../DataDefine/DamageType";
import DungeonManager from "../../../../DungeonManager";
import { OnPreBattleMessage } from "../../../../Events/OnPreBattleMessage";
import { LocalizationProperty } from "../../../../Gen/DB/Localization";
import { SporeCloudPower } from "../../../../Power/SporeCloudPower";
import { StrengthPower } from "../../../../Power/StrengthPower";
import { AbstractMonster } from "../../AbstractMonster";
import { Intent } from "../../Intent";
import { MonsterInfo } from "../../MonsterInfo";

export class Cultist_ViewModel extends AbstractMonster{
    private StrAmt;
    public constructor(monster:MonsterInfo) {
        super(monster);
    }
    Initialize() {
        if(DungeonManager.Inst.AdvanceLevel >= 2) {
            this.StrAmt = 5;
            this.DamageInfo.push(new DamageInfo(this,6));
        }else{
            this.StrAmt = 3;
            this.DamageInfo.push(new DamageInfo(this,6));
        }
        let trackEntry = <Spine.TrackEntry>this.SetAnimation("Idle",true);
        trackEntry.TrackTime = trackEntry.TrackEnd * Mathf.Random()
        trackEntry.TimeScale = Mathf.RandomRange(0.7,1);
    }


    PreBattleAction(msg: OnPreBattleMessage) {
        super.PreBattleAction(msg);
        this.AddToBot(new ApplyPowerAction(this,this,new SporeCloudPower(2)))
    }

    GetMove(num: number) {
        if(num <60){
            if(this.LastTwoMoves(1)){
                this.SetMove(2,Intent.BUFF,{moveName:LocalizationProperty.Read("咔咔")})
            }
        }
        if(DungeonManager.Inst.CurrentRoom.Round == 1){
            this.SetMove(3,Intent.BUFF,{moveName:LocalizationProperty.Read("念咒")})
        }
        this.SetMove(1,Intent.ATTACK,{damage:this.DamageInfo[0]});
    }

    TakeTurn() {
        switch(this.MoveIndex){
            case 1:{
                this.AddToBot(new DoFuncAction(()=>this.ChangeState("ATTACK"),0.5));
                this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0],AttackEffect.BLUNT_LIGHT));
                break;
            }
            case 2:{
                this.AddToBot(new ApplyPowerAction(this,this,new StrengthPower(),this.StrAmt));
                break;
            }
        }
        this.AddToBot(new RollMoveAction(this));
    }
    
    private ChangeState(key:string){
        switch(key){
            case "ATTACK":{
                let animation = <Spine.Unity.SkeletonAnimation>this.DisplayObject.Self;
                animation.AnimationState.SetAnimation(0,"ATTACK",false);
                animation.AnimationState.AddAnimation(0,"Idle",true,0);
                break;
            }
        }
    }

    Damage(info: DamageInfo) {
        super.Damage(info);
        if(info.DamageType == DamageType.THORNS && info.TotalDamage > 0){
            let animation = <Spine.Unity.SkeletonAnimation>this.DisplayObject.Self;
            animation.AnimationState.SetAnimation(0,"Hit",false);
            animation.AnimationState.AddAnimation(0,"Idle",true,0);
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/ViewModel/TheBottom/FungiBeast_ViewModel.cs
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using ModTheSpire.Power.Buff;
using Spine;
using Spine.Unity;
using TCG.Config;
using UnityEngine;

namespace ModTheSpire.Unit.Monster.ViewModel.TheBottom
{
    public class FungiBeast_ViewModel : AbstractMonster
    {
        private int mStrAmt;
        protected override void Initialize()
        {
            mStrAmt = DungeonManager.Inst.AdvanceLevel >= 2 ? 5 : 3;
            DamageInfo.Add(new DamageInfo(this,6));
            var trackEntry = (TrackEntry)this.DisplayObject.SetAnimation("idle", true);
            trackEntry.TrackTime = trackEntry.TrackEnd * Random.value;
            trackEntry.TimeScale = Random.Range(0.7f,1);
        }

        protected override void PreBattleAction(OnPreBattleMessage msg)
        {
            base.PreBattleAction(msg);
            AddToBot(new ApplyPowerAction(this, this, new SporeCloudPower(), 2));
        }

        protected override void GetMove(int num)
        {
            if (num < 60) {
                if (this.LastTwoMoves(1)) {
                    this.SetMove(2, Intent.BUFF,moveName:LocalizationProperty.Read("生长"));
                } else {
                    this.SetMove(1, Intent.ATTACK, this.DamageInfo[0]);
                }
            } else if (this.LastMove(2))
            {
                this.SetMove(1, Intent.ATTACK, this.DamageInfo[0]);
            } else {
                this.SetMove(2, Intent.BUFF,moveName:LocalizationProperty.Read("生长"));
            }
        }

        protected override void TakeTurn()
        {
            switch(this.MoveIndex){
                case 1:{
                    this.AddToBot(new DoFuncAction(()=>this.ChangeState("ATTACK"),500));
                    this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0],AttackEffect.BLUNT_LIGHT));
                    break;
                }
                case 2:{
                    this.AddToBot(new ApplyPowerAction(this,this,new StrengthPower(),this.mStrAmt));
                    break;
                }
            }
            this.AddToBot(new RollMoveAction(this));
        }
        
        private void ChangeState(string key){
            switch(key){
                case "ATTACK":{
                    var animation = (SkeletonAnimation)this.DisplayObject.Self;
                    animation.AnimationState.SetAnimation(0,"ATTACK",false);
                    animation.AnimationState.AddAnimation(0,"Idle",true,0);
                    break;
                }
            }
        }

        public override void Damage(DamageInfo info)
        {
            base.Damage(info);
            if(info.Type == DamageType.THORNS && info.Total > 0){
                var animation = (SkeletonAnimation)this.DisplayObject.Self;
                animation.AnimationState.SetAnimation(0,"Hit",false);
                animation.AnimationState.AddAnimation(0,"Idle",true,0);
            }
        }

        public FungiBeast_ViewModel(MonsterInfo info) : base(info)
        {
        }
    }
}